### PR TITLE
[cli][test] validation of schemas without writing a rule

### PR DESCRIPTION
**Pull request by ryandeivert**
to @austinbyers for manage.py updates and various logging changes
to @jacknagz and @chunyong-lin for cli refactor
to @mime-frame: please confirm that log output examples below are sensible
size: medium
resolves: #149 
resolves: #154 
resolves: #245
contributes to: #276 

### Changes
* Testing via `manage.py` now supports a `validate-schemas` command.
  * An optional argument for this command is `--test-files` which accepts a space separated list of files (or, alternatively, rule names) to validate.
  * For example: `python manage.py validate-schemas --test-files cloudtrail_put_bucket_acl`
  * This command does not perform any actual rule engine operations and will simply attempt to 'classify' the record against the defined log schemas.
* All summary output is now printed through the logging object with their respective levels (ie: warning, error, info) instead of through print statements.

##### `$ python manage.py validate-schemas --test-files cloudtrail_put_object_acl.json`
```
$ python manage.py validate-schemas --test-files cloudtrail_put_object_acl.json
StreamAlertCLI [INFO]: Issues? Report here: https://github.com/airbnb/streamalert/issues

cloudtrail_put_object_acl
       [Pass]  [log='cloudtrail:events']     validation  (s3): CloudTrail - PutObjectAcl - True Positive
       [Pass]  [log='cloudtrail:events']     validation  (s3): CloudTrail - PutObjectAcl - False Positive

StreamAlertCLI [INFO]: (2/2) Successful Tests
StreamAlertCLI [INFO]: Completed
```

##### `$ python manage.py validate-schemas --test-files cloudtrail_put_object_acl.json`
```
$ python manage.py validate-schemas --test-files cloudtrail_put_object_acl.json
StreamAlertCLI [INFO]: Issues? Report here: https://github.com/airbnb/streamalert/issues

cloudtrail_put_object_acl
       [Pass]  [log='cloudtrail:events']     validation  (s3): CloudTrail - PutObjectAcl - True Positive
       [Fail]  [log='unknown']               validation  (s3): CloudTrail - PutObjectAcl - False Positive

StreamAlertCLI [INFO]: (1/2) Successful Tests
StreamAlertCLI [ERROR]: (1/2) Failures
StreamAlertCLI [ERROR]: (1/1) [cloudtrail_put_object_acl] Data is invalid due to missing key(s) in test record: 'eventVersion'. Rule: 'cloudtrail_put_object_acl'. Description: 'CloudTrail - PutObjectAcl - False Positive'
```

### Other updates
* Adding the ability to detect any error messages that are logged during execution
  * By adding a MemoryHandler to the root logger and filtering out all messages that are not of level 'error', we can store any errors that occur during test runs in a buffer.
  * The number of errors that occurred will be reported at the end, and the errors themselves will be logged so they can be addressed.
* We will now log warnings to stdout for any rule that exists without a corresponding integration test, like so:
```
StreamAlertCLI [WARNING]: No tests configured for rule: 'rule_without_a_test'. Please add a corresponding test file for this rule in 'tests/integration/rules' with the name 'rule_without_a_test.json' to avoid seeing this warning
```
* A similar warning like the one above will be logged if a integration test file exists, but there is no rule for said file. However, this will **only** occur if the command is `lambda test` and not with `validate-schemas`. The output is like so:
```
StreamAlertCLI [WARNING]: No rules configured for test file: 'test_without_rule.json'. Please add a corresponding rule for this test file in 'rules/' with the name 'test_without_rule.py' to avoid seeing this warning and any associated errors above
```
* Significantly changing how status messages are stored and output during a test run.

### Bug fixes
* Resolving bug with improper set logic that could cause the alert processor to get ran inadvertently. See [this commit](https://github.com/airbnb/streamalert/commit/817d9e348300fd718498ab0adb67cfed0070fde6).
* Resolving bug that could cause the `analyze_record_delta` function to report that keys defined as `optional_top_level_keys` in a log's schema were in fact required. See [this commit](https://github.com/airbnb/streamalert/commit/c9c493579403108a31bf6292143fcab81c41c741).

### CODE REVIEW

File: stream_alert_cli/test.py [hunk 1]
```diff
@@ -86,81 +86,132 @@ def __init__(self, context, print_output):
         self.print_output = print_output
         self.invalid_log_messages = []
 
-    def test_processor(self, rules):
+    def test_processor(self, filter_rules):
         """Perform integration tests for the 'rule' Lambda function
 
         Args:
-            rules (list|None): Specific rule names (or None) to restrict
+            filter_rules [list or None]: Specific rule names (or None) to restrict
```
Comment (jacknagz) [inline stream_alert_cli/test.py]: Update the docstrings to use `(type)` per our new convention pls

File: stream_alert_cli/test.py [hunk 2]
```diff
@@ -86,81 +86,132 @@ def __init__(self, context, print_output):
...
-                message = ('' if current_test_passed else
-                           'Rule failure: {}'.format(test_record['description']))
-                self.status_messages.append(StatusMessage(message_type, rule_name, message))
+                yield self._run_rule_tests(rule_name, test_record, print_header_line)
 
         # Report on the final test results
         self.report_output_summary()
 
+    def _run_rule_tests(self, rule_name, test_record, print_header_line):
+        """Run tests on a test record for a given rule
```
Comment (jacknagz) [inline stream_alert_cli/test.py]: take another look at this docstring description

File: stream_alert_cli/test.py [hunk 3]
```diff
@@ -86,81 +86,132 @@ def __init__(self, context, print_output):
...
+            [generator] Yields back the rule name and the json loaded contents of
+                the respective test event file.
+        """
+        for _, _, test_rule_files in os.walk(DIR_RULES):
+            for rule_file in test_rule_files:
+                rule_name = rule_file.split('.')[0]
+
+                # If only specific rules are being tested,
+                # skip files that do not match those rules
+                if filter_rules:
```
Comment (jacknagz) [inline stream_alert_cli/test.py]: chain these two `if`s together

File: manage.py [hunk 4]
```diff
@@ -140,6 +140,48 @@ def _add_live_test_subparser(subparsers):
     )
 
 
+def _add_validate_schema_subparser(subparsers):
```
Comment (jacknagz) [inline manage.py]: pylint doesn't complain about no docstring?

File: manage.py [hunk 5]
```diff
@@ -140,6 +140,48 @@ def _add_live_test_subparser(subparsers):
...
+    schema_validation_usage = 'manage.py validate-schemas [options]'
+    schema_validation_description = ("""
+StreamAlertCLI v{}
+Run end-to-end tests that will attempt to send alerts
+
+Available Options:
+
+    --files                 Name (not full path) of test file(s) to validate, separated by spaces
```
Comment (jacknagz) [inline manage.py]: Should we call this `--test-files`? I'd also add a little description of what these files should contain

File: manage.py (hunk 4 above)
Comment (ryandeivert) [inline manage.py]: Non-public mehods/functions don't actually _need_ docstrings so it wouldn't complain here.. but I didn't realize this was missing one so I'll add it for consistency

File: manage.py (hunk 5 above)
Comment (ryandeivert) [inline manage.py]: Sure that's a good change - I'll update the comment to clarify some stuff

**Comment by ryandeivert** [comment]
@jacknagz  PTAL at new commits - one for your feedback and one for doc updates

**Comment by mime-frame** [comment]
It took me awhile to understand what exactly was being passed to `--files` in this example:

`$ python manage.py validate-schemas --files cloudtrail_put_object_acl`

Proposal: We change it from `--files` to `--log-files`

**Comment by austinbyers** [comment]
@ryandeivert Just a note that you can't actually use a comma-separated list for "Resolves" in the PR description :( If you hover over it, it only shows that it will resolve the first listed issue

**Comment by mime-frame** [comment]
I'm a bit confused. We state this:

> This command does not perform any actual rule engine operations and will simply attempt to 'classify' the record against the defined log schemas.

Yet, the output references rule integration tests and whether they passed?:

```
cloudtrail_put_object_acl
       [Pass]  [log='cloudtrail:events']     validation  (s3): CloudTrail - PutObjectAcl - True Positive
       [Pass]  [log='cloudtrail:events']     validation  (s3): CloudTrail - PutObjectAcl - False Positive
```

File: manage.py (hunk 4 above)
Comment (austinbyers) [inline manage.py]: Good to know

File: manage.py [hunk 6]
```diff
@@ -140,6 +140,56 @@ def _add_live_test_subparser(subparsers):
...
+def _add_validate_schema_subparser(subparsers):
+    """Add the validate-schemas subparser: manage.py validate-schemas [options]"""
+    schema_validation_usage = 'manage.py validate-schemas [options]'
+    schema_validation_description = ("""
+StreamAlertCLI v{}
+Run end-to-end tests that will attempt to send alerts
+
+Available Options:
+
+    --test-files         Name(s) of test files to validate, separated by spaces (not full path)
```
Comment (austinbyers) [inline manage.py]: Isn't it usually the standard to separate multiple arguments in CLI by commas instead of spaces? Each space-delimited token should be its own separate argument

File: stream_alert_cli/helpers.py [hunk 7]
```diff
@@ -23,6 +24,7 @@
 import zlib
 
 import boto3
+from moto import mock_cloudwatch, mock_lambda, mock_kms, mock_s3
```
Comment (austinbyers) [inline stream_alert_cli/helpers.py]: Nit: alphabetical import `mock_kms` before `mock_lambda`

File: stream_alert_cli/logger.py [hunk 8]
```diff
@@ -33,3 +54,14 @@
...
+# Get a logging MemoryHandler with a default buffer size of 1000
+# We don't care about assigning a target to this handler since these logs
+# will not actually be written out to disk, etc
+LOG_ERROR_HANDLER = logging.handlers.MemoryHandler(1000)
+
+# Add a filter to suppress everything that is not an error
+LOG_ERROR_HANDLER.addFilter(SuppressNonErrors())
+
+# Add the MemoryHandler to the root logger to capture all logs in all loggers
+logging.getLogger().addHandler(LOG_ERROR_HANDLER)
```
Comment (austinbyers) [inline stream_alert_cli/logger.py]: If this extra log handler is only used for unit tests, maybe we should only add it if the user is running tests? Instead of adding an extra in-memory logging handler for all StreamAlert operations

Comment (ryandeivert) [inline stream_alert_cli/logger.py]: This isn't unit tests and doesn't get added for all streamalert operations - this happens only in the CLI package - but you are right I can update to only happen if integration tests are being run! making that change

**Comment by ryandeivert** [comment]
@mime-frame idk how you commented but I am not able to reply directly to you. re this comment: https://github.com/airbnb/streamalert/pull/277#issuecomment-324720347

I think your preexisting knowledge of output from other commands (`python manage.py lambda test ...`) may be skewing how you are interpreting this output. The `Pass` and `Fail` in this case is related to matching a defined schema (not triggering a rule)..

For instance, we print `[Pass]  [log='cloudtrail:events']     validation ...` meaning "this validation test passed, and the log type matched to is `cloudtrail:events`"

The alternative was to print nothing at all :)

Also RE this comment: https://github.com/airbnb/streamalert/pull/277#issuecomment-324719919  there was a new commit pushed a few hours ago that changes this to `--test-files`

**Comment by mime-frame** [comment]
Cool, thanks @ryandeivert - my review is complete

File: manage.py (hunk 6 above)
Comment (ryandeivert) [inline manage.py]: Yes I agree with you here, but argparse doesn't :) List supplied as args to argparse are expected to be space delimited. We could change this to expect a string that is a comma separated and handle all of the logic in code, but then we lose the validation/err handling baked into argparse. At one point (with the --rules flag) I was doing comma separated but we decided to do spaces since it was safer with less overhead

Comment (austinbyers) [inline manage.py]: Oh interesting. Ok, sounds good! Thanks for the note

File: manage.py [hunk 9]
```diff
@@ -140,6 +140,56 @@ def _add_live_test_subparser(subparsers):
...
+    schema_validation_parser.set_defaults(command='validate-schemas')
+
+    # add the optional ability to test against specific files
+    schema_validation_parser.add_argument(
+        '-f', '--test-files',
+        nargs='+',
+        help=ARGPARSE_SUPPRESS
+    )
+
+    # allow verbose output for the CLI with te --debug option
```
Comment (austinbyers) [inline manage.py]: nit: typo (`with te --debug option`)

File: stream_alert_cli/helpers.py [hunk 10]
```diff
@@ -203,3 +205,68 @@ def put_mock_s3_object(bucket, key, data, region):
...
+
+    Args:
+        context (namedtuple): A constructed aws context object
+    """
+    def wrap(func):
+        """Wrap the returned function with or without mocks"""
+        if context.mocked:
+            @mock_cloudwatch
+            @mock_lambda
+            @mock_s3
```
Comment (austinbyers) [inline stream_alert_cli/helpers.py]: Just beware of https://github.com/spulec/moto/issues/1026; the mock may not be restored correctly every time

File: stream_alert_cli/test.py [hunk 11]
```diff
@@ -68,98 +71,223 @@ def __init__(self, context, print_output):
...
 
-    def test_processor(self, rules):
+    def test_processor(self, filter_rules, validate_only=False):
         """Perform integration tests for the 'rule' Lambda function
 
         Args:
-            rules (list|None): Specific rule names (or None) to restrict
+            filter_rules (list|None): Specific rule names (or None) to restrict
                 testing to. This is passed in from the CLI using the --rules option.
+            validate_only [bool=False]: If true, validation of test records will occur
```
Comment (austinbyers) [inline stream_alert_cli/test.py]: Nit: should be `validate_only (bool)`. Types should be in parens (not brackets), and the default value need not be specified since it's in the function definition

File: docs/source/rule-testing.rst [hunk 12]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
   }
 
 
+Validating Log Schemas
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: `Validate Log Schemas`

File: docs/source/rule-testing.rst [hunk 13]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+~~~~~~~~~~~~~~~~~~~~~~
+
+In some cases, there may be logs coming into StreamAlert that are of a known type, but you do not have a current need to apply rule
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: `may be incoming logs to StreamAlert with a known type but without specific rules`

File: docs/source/rule-testing.rst [hunk 14]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+logic to them. However, it is still advantageous to add schemas for these logs and *test* to ensure your new schemas are valid.
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: `However, it is best practice to write schemas for these logs and *verify* that they are valid.`

File: docs/source/rule-testing.rst [hunk 15]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+
+This is possible through the addition of log schema(s) to ``conf/logs.json`` and creation of test record(s) in ``tests/integration/rules/``
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: `This is possible by first adding the new schema(s) to ``conf/logs.json`` along with creation of test record(s)...`

File: docs/source/rule-testing.rst [hunk 16]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+using known samples of logs (without actually adding a corresponding rule). Running the ``manage.py`` script with the ``validate-schemas``
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: `containing samples of real logs.`

File: docs/source/rule-testing.rst [hunk 17]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+option will iterate over all json test files and attempt to match records within them to a defined log schema.
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: I think it's cool to use `classify records` here

File: docs/source/rule-testing.rst [hunk 18]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+
+To simply run schema validation on all test files:
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: nix `simply`

File: docs/source/rule-testing.rst [hunk 19]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+  $ python manage.py validate-schemas --test-files <test_rule_file.json>
+
+Or:
+
+.. code-block:: bash
+
+  $ python manage.py validate-schemas --test-files <test_rule_file>
+
+
+Example of running schema validation on two valid test files:
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: `s/Example of running//`

File: docs/source/rule-testing.rst [hunk 20]
```diff
@@ -76,6 +76,84 @@ For example, to replace a time based field with ``last_hour``:
...
+         [Pass]  [log='cloudwatch:events']     validation  (kinesis): CloudTrail - PutBucketAcl - True Positive
+         [Pass]  [log='cloudwatch:events']     validation  (kinesis): CloudTrail - PutBucketAcl - False Positive
+
+
+
+  StreamAlertCLI [INFO]: (10/10) Successful Tests
+  StreamAlertCLI [INFO]: Completed
+
+
+Example of running schema validation on a test file that contains one valid record and one invalid record:
```
Comment (jacknagz) [inline docs/source/rule-testing.rst]: maybe shorten this to `failed record` or `failed test` example?

File: manage.py (hunk 9 above)
Comment (ryandeivert) [inline manage.py]: good catch!

File: stream_alert_cli/helpers.py (hunk 10 above)
Comment (ryandeivert) [inline stream_alert_cli/helpers.py]: These tests don't actually run on deploy/in production but good to keep in mind - thanks!

File: docs/source/rule-testing.rst (hunk 19 above)
Comment (ryandeivert) [inline docs/source/rule-testing.rst]: unclear what you're looking for here? just `Schema validation on two valid test files:`?

File: docs/source/rule-testing.rst (hunk 20 above)
Comment (ryandeivert) [inline docs/source/rule-testing.rst]: I think explaining what is actually in the output makes it more obvious to the user what is happening. Saying `Failed test` doesn't really explain what failed very well
